### PR TITLE
CSharp - Update for conan V2 and build for 3.8.0 (+ add arm64)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

**Comment by jmarrec** [comment]
Gonna close this PR temporarily so I don't get the cppcheck/clangformat running for no reason on it

**Comment by jmarrec** [comment]
Alright 3.8.0 packages, including the new osx-arm64 are available on this run: https://github.com/NREL/OpenStudio/actions/runs/9189117111/job/25275666930 

@kbenne I hosted the windows (that contains both win32 and x64) on https://www.nuget.org/packages/OpenStudio/3.8.0

@MingboPeng FYI if you want to upload them to your nuget packages per platform.

File: conan.lock [hunk 1]
```diff
@@ -5,7 +5,7 @@
         "websocketpp/0.8.2#842a0419153a8aa52f3ea3a1da557d38%1695972005.713",
         "tinygltf/2.5.0#65c28d0a4c3cbd4ef92b08b59df769da%1701621757.442",
         "termcap/1.3.1#1986f84bf21dd07ea774b027a3201fcb%1678542508.75",
-        "swig/4.1.1#8f418438101ef7712d52a9fe739d32f9%1707838298.077707",
+        "swig/4.1.1#2bb5c79321cbb05bcab525c690d9bf74%1716336914.081294",
```
Comment (jmarrec) [inline conan.lock:8]: Tweaked and built from https://github.com/jmarrec/conan-center-index/commit/52e12ba5e491462dfb4bba5e86c4c141db2735e9

File: conanfile.py [hunk 2]
```diff
@@ -15,16 +15,23 @@ class OpenStudioBuildRecipe(ConanFile):
...
+        "with_ruby": [True, False],
+        "with_python": [True, False],
+        "with_csharp": [True, False],
         "with_testing": [True, False],
         "with_benchmark": [True, False],
     }
     default_options = {
+        "with_ruby": True,
+        "with_python": True,
+        "with_csharp": False,
```
Comment (jmarrec) [inline conanfile.py:27]: New options to allow disabling ruby (won't build on win32), and tweak the cmake vars as well.

File: src/measure/Measure.i [hunk 3]
```diff
@@ -24,6 +24,15 @@
   // Help in overload resolution preferring std::string over char const *
   %ignore openstudio::measure::OSArgument::setValue(char const*);
   %ignore openstudio::measure::OSArgument::setDefaultValue(char const*);
+
+  // Ignore Json::Value return type (toJSON / fromJSON / valueAsJSON are globally ignored already)
+  %ignore openstudio::measure::OSArgument::defaultValueAsJSON;
+  %ignore openstudio::measure::OSArgument::domainAsJSON;
+  %ignore openstudio::measure::OSRunner::getArgumentValues;
+  %ignore openstudio::measure::OSRunner::getPastStepValuesForMeasure;
+  %ignore openstudio::measure::OSRunner::getPastStepValuesForName;
```
Comment (jmarrec) [inline src/measure/Measure.i:33]: Ignore new Json::Value returns

File: csharp/OpenStudio-osx-arm64.targets
Comment (jmarrec) [inline csharp/OpenStudio-osx-arm64.targets:1]: New

File: csharp/developer/OpenStudio/OpenStudio.csproj.in [hunk 4]
```diff
@@ -58,25 +58,45 @@ openstudiolib.dll
...
+    <ItemGroup Condition=" $(IsDarwin) And !$(IsArm64) ">
       <TfmSpecificPackageFile Include="../Products/*openstudio*csharp.dylib" PackagePath="runtimes/osx-x64/native" />
       <TfmSpecificPackageFile Include="../Products/*openstudiolib.dylib" PackagePath="runtimes/osx-x64/native" />
       <TfmSpecificPackageFile Include="../Products/OpenStudio-osx-x64.targets" PackagePath="build/OpenStudio.targets" />
     </ItemGroup>
+    <ItemGroup Condition=" $(IsDarwin) And $(IsArm64) ">
+      <TfmSpecificPackageFile Include="../Products/*openstudio*csharp.dylib" PackagePath="runtimes/$(Rid)/native" />
+      <TfmSpecificPackageFile Include="../Products/*openstudiolib.dylib" PackagePath="runtimes/$(Rid)/native" />
+      <TfmSpecificPackageFile Include="../Products/OpenStudio-$(Rid).targets" PackagePath="build/OpenStudio.targets" />
+    </ItemGroup>
```
Comment (jmarrec) [inline csharp/developer/OpenStudio/OpenStudio.csproj.in:98]: New to support osx-arm64

File: csharp/developer/OpenStudio/OpenStudio.csproj.in [hunk 5]
```diff
@@ -58,25 +58,45 @@ openstudiolib.dll
...
 
     <!-- for Mac  -->
-    <ItemGroup Condition=" '$([System.Runtime.InteropServices.RuntimeInformation]::IsOSPlatform($([System.Runtime.InteropServices.OSPlatform]::OSX)))' ">
+    <ItemGroup Condition=" $(IsDarwin) And !$(IsArm64) ">
       <TfmSpecificPackageFile Include="../Products/*openstudio*csharp.dylib" PackagePath="runtimes/osx-x64/native" />
       <TfmSpecificPackageFile Include="../Products/*openstudiolib.dylib" PackagePath="runtimes/osx-x64/native" />
       <TfmSpecificPackageFile Include="../Products/OpenStudio-osx-x64.targets" PackagePath="build/OpenStudio.targets" />
     </ItemGroup>
+    <ItemGroup Condition=" $(IsDarwin) And $(IsArm64) ">
+      <TfmSpecificPackageFile Include="../Products/*openstudio*csharp.dylib" PackagePath="runtimes/$(Rid)/native" />
```
Comment (jmarrec) [inline csharp/developer/OpenStudio/OpenStudio.csproj.in:95]: Note that for mac, I could just use a single block... But I was wary of maybe breaking something, and after about 25 runs of the C# workflow (some of which I have to wait almost 2 hours to see if breaks), I just didn't want to risk it

```
    <ItemGroup Condition=" $(IsDarwin) ">
      <TfmSpecificPackageFile Include="../Products/*openstudio*csharp.dylib" PackagePath="runtimes/$(Rid)/native" />
      <TfmSpecificPackageFile Include="../Products/*openstudiolib.dylib" PackagePath="runtimes/$(Rid)/native" />
      <TfmSpecificPackageFile Include="../Products/OpenStudio-$(Rid).targets" PackagePath="build/OpenStudio.targets" />
    </ItemGroup>
```

File: src/utilities/core/jsoncpp.i [hunk 6]
```diff
@@ -5,6 +5,7 @@
   // ignores toJSON/fromJSON globally
   %rename("$ignore", regextarget=1, fullname=1) "openstudio::.*::toJSON$";
   %rename("$ignore", regextarget=1, fullname=1) "openstudio::.*::fromJSON$";
+  %rename("$ignore", regextarget=1, fullname=1) "openstudio::.*::valueAsJSON$";
```
Comment (jmarrec) [inline src/utilities/core/jsoncpp.i:8]: New global ignore

**Comment by MingboPeng** [comment]
> Alright 3.8.0 packages, including the new osx-arm64 are available on this run: https://github.com/NREL/OpenStudio/actions/runs/9189117111/job/25275666930
> 
> @kbenne I hosted the windows (that contains both win32 and x64) on https://www.nuget.org/packages/OpenStudio/3.8.0
> 
> @MingboPeng FYI if you want to upload them to your nuget packages per platform.

Thanks @jmarrec, done! The new NuGet packages are added per platform including arm64 for mac.

**Pull request merged**